### PR TITLE
feat(query,prometheus): Extract histogram bucket with _bucket_ magic filter

### PR DESCRIPTION
**Pull request by velvia**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [x] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** (link exiting issues here : https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)

Currently, our `histogram_bucket` function is the only way to extract a bucket out of FiloDB histograms.  However, the function syntax does not combine well with other operators and range functions.  For example, one cannot do `rate(histogram_bucket(2.5, my_histogram{.....}[5m]))` right now.  The syntax also seems strange to someone coming from pure Prometheus.

**New behavior :**

Introduce a magic label filter, `_bucket_`.  Thus the above query can be rewritten thus: `rate(my_histogram{...., _bucket_="2.5"}[5m])`.... ie Prometheus users just need to change `_le_` to `_bucket_`.

The above query is translated into a LogicalPlan where an `ApplyInstantFunction` is done below the `PeriodicSeriesWithWindowing`.  Instant function is then applied to the raw histogram to extract out the buckets.

NOTE: this is a partial PR.   A few minor changes are still needed in order for an ExecPlan to be properly translated from the logical plan and executed properly.  Sending this out early for conceptual and PromQL feedback.

### CODE REVIEW

File: doc/query-engine.md [hunk 1]
```diff
@@ -205,6 +203,31 @@ E~BinaryJoinExec(binaryOp=DIV, on=List(), ignoring=List()) on ActorPlanDispatche
 See [QueryEngineSpec](../coordinator/src/test/scala/filodb.coordinator/queryengine2/QueryEngineSpec.scala) for
 test code on this example. 
 
+## Special Case: Extracting Buckets Out of Histograms
+
+FiloDB stores all histogram buckets together as a single time series, leading to huge performance and compression advantages.  Due to this, however, extracting a single bucket out of a histogram is much more involved than simply selecting a time series like Prometheus using a `le=XXX` filter.  For FiloDB, it has to involve transforming the source histogram into a single bucket, using the function `histogram_bucket`. 
+
+Let's say one wanted to plot a single bucket which represents count of all latencies below 500ms:
+
+    rate(http_req_latency_buckets{app="foobar",le=".500"}[5m])
```
Comment (beorn7) [inline doc/query-engine.md]: Nit: `rate(http_req_latency_bucket{app="foobar",le=".5"}[5m])`

Comment (velvia) [reply doc/query-engine.md]: 👍 thanks

File: prometheus/src/main/scala/filodb/prometheus/ast/Vectors.scala [hunk 2]
```diff
@@ -110,14 +113,24 @@ trait Vectors extends Scalars with TimeUnits with Base {
...
      */
     protected def labelMatchesToFilters(labels: Seq[LabelMatch]) = {
       var column: Option[String] = None
-      val filters = labels.map { labelMatch =>
+      var bucketOpt: Option[Double] = None
+      val filters = labels.filter { labelMatch =>
+        if (labelMatch.label == BucketFilterLabel) {
+          bucketOpt = parseBucketValue(labelMatch.value)
+          false
+        } else true
```
Comment (vishramachandran) [inline prometheus/src/main/scala/filodb/prometheus/ast/Vectors.scala]: Assuming `_type_` is going to be handled here as well ?

File: prometheus/src/main/scala/filodb/prometheus/ast/Vectors.scala [hunk 3]
```diff
@@ -154,17 +167,22 @@ trait Vectors extends Scalars with TimeUnits with Base {
...
-      PeriodicSeries(
+      val ps = PeriodicSeries(
         RawSeries(timeParamToSelector(timeParams, staleDataLookbackSeconds * 1000),
           columnFilters, column.toSeq),
         timeParams.start * 1000, timeParams.step * 1000, timeParams.end * 1000
       )
+      bucketOpt.map { bOpt =>
+        // It's a fixed value, the range params don't matter at all
+        val param = ScalarFixedDoublePlan(bOpt, RangeParams(0, Long.MaxValue, 60000L))
+        ApplyInstantFunction(ps, InstantFunctionId.HistogramBucket, Seq(param))
```
Comment (vishramachandran) [inline prometheus/src/main/scala/filodb/prometheus/ast/Vectors.scala]: Now that I have had a chance to see this PR fully and understand some of the changes that come later on SeriesPlan, I am of the opinion that Histogram Bucket extraction shouldnt be an instant function at the logical plan.

We are including some non-use cases in the logical plan in order to include a corner use case. For example `sum(log(RawSeriesPlan))` (shortened for brevity). Let us discuss offline in more detail.

Comment (velvia) [reply prometheus/src/main/scala/filodb/prometheus/ast/Vectors.scala]: Resolved through offline discussion.... will push changes to make LogicalPlans more type safe.

File: prometheus/src/main/scala/filodb/prometheus/ast/Vectors.scala (hunk 2 above)
Comment (velvia) [reply prometheus/src/main/scala/filodb/prometheus/ast/Vectors.scala]: It is handled in materialization of LPs.  Can address and improve that separately.

File: query/src/main/scala/filodb/query/LogicalPlan.scala [hunk 4]
```diff
@@ -94,7 +97,7 @@ case class PeriodicSeries(rawSeries: RawSeriesPlan,
   * Applies a range function on raw windowed data before
   * sampling data at regular intervals.
   */
-case class PeriodicSeriesWithWindowing(rawSeries: RawSeries,
+case class PeriodicSeriesWithWindowing(rawSeries: SeriesPlan,
```
[the remark this answers is not in the document]
Comment (velvia) [reply query/src/main/scala/filodb/query/LogicalPlan.scala]: See the pushed changes

File: query/src/main/scala/filodb/query/LogicalPlan.scala [hunk 5]
```diff
@@ -91,18 +94,18 @@ case class PeriodicSeries(rawSeries: RawSeriesPlan,
   * Concrete logical plan to query for data in a given range
   * with results in a regular time interval.
   *
-  * Applies a range function on raw windowed data before
+  * Applies a range function on raw windowed data (perhaps with instant function applied) before
   * sampling data at regular intervals.
   */
-case class PeriodicSeriesWithWindowing(rawSeries: RawSeries,
+case class PeriodicSeriesWithWindowing(series: NonPeriodicSeriesPlan,
```
Comment (vishramachandran) [inline query/src/main/scala/filodb/query/LogicalPlan.scala]: should be RawSeries. No need for NonPeriodicSeriesPlan

File: query/src/main/scala/filodb/query/LogicalPlan.scala [hunk 6]
```diff
@@ -2,17 +2,20 @@ package filodb.query
...
 
+//scalastyle:off number.of.types
 sealed trait LogicalPlan {
   /**
     * Execute failure routing
     */
   def isRoutable: Boolean = true
 }
 
+sealed trait NonPeriodicSeriesPlan extends LogicalPlan
```
[the remark this answers is not in the document]
Comment (velvia) [reply query/src/main/scala/filodb/query/LogicalPlan.scala]: Removed.

**Comment by velvia** [comment]
@vishramachandran  have a look again, rebased and updated names and removed the `NonPeriodicPlan` trait.

**Pull request merged**